[FLINK-4988] [elasticsearch] Add Elasticsearch 5.x Connector

tzulitai opened the pull request:
This PR is based on @mikedias's work in #2767 (1st commit), with additional work to restructure the Elasticsearch connectors and documentation (2nd commit). Basically, we now have a `flink-connector-elasticsearch-base` containing common behaviour and test code, and all version-specific modules stem from that.

### Deprecated Constructors for Elasticsearch 1.x / 2.x
As part of the restructuring, all connector versions now take a `ElasticsearchSinkFunction` (previously, 1.x took a `IndexRequestBuilder`, which was limited to only indexing actions on a Elasticsearch index) for full functional Elasticsearch support.

The `ElasticsearchSinkFunction` is therefore reallocated from package `o.a.f.s.c.elasticsearch2` in `flink-connector-elasticsearch2`, to `o.a.f.s.c.elasticsearch` in `flink-connector-elasticsearch-base`.

This resulted in deprecation of the following:

1. In Elasticsearch 1.x: All original `IndexRequestBuilder` constructors as well as the interface itself have been deprecated.
2. In Elasticsearch 2.x: Due to the package relocation of `ElasticsearchSinkFunction`, all original constructors are also deprecated in favor of the new package path for the class.


R: @rmetzger @mikedias please feel free to review, thank you!

rmetzger commented:
Thanks a lot for opening a pull request for this.
It looks like some of the tests are failing on travis. Does ES5 support Java 7 ?

rmetzger commented on docs/dev/connectors/elasticsearch.md:
````diff
@@ -23,158 +23,284 @@ specific language governing permissions and limitations
+<transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer">
+   <resource>META-INF/services/org.apache.lucene.codecs.PostingsFormat</resource>
+</transformer>
+```
````
+1 for explaining how to resolve this issue!
It would be good if you could mention in the docs where to add the transformers exactly.
Also, I'm wondering if you could use the ServicesTransformer instead: http://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer

tzulitai replied:
I'll try it out!

rmetzger commented on docs/dev/connectors/elasticsearch2.md:
```diff
@@ -1,173 +0,0 @@
----
```
Can you replace the es2 page with a redirect to the new page?
This way existing links are not broken.

tzulitai replied:
Will do!

rmetzger commented on docs/dev/connectors/elasticsearch2.md:
```diff
@@ -1,173 +0,0 @@
-</transformer>
-<transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer">
-   <resource>META-INF/services/org.apache.lucene.codecs.PostingsFormat</resource>
-</transformer>
```
Ah, I see. This has been there before :) Still, would be great if you could fix it :)

tzulitai commented:
@rmetzger 
Ah ... seems like ES 5.x requires at least Java 8.
- https://www.elastic.co/guide/en/elasticsearch/reference/master/_installation.html#_installation
- https://github.com/elastic/elasticsearch/issues/17584

rmetzger commented:
Using maven build profiles, you can probably include the es5 connector in java8 builds only.

rmetzger commented on flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java:
```diff
@@ -0,0 +1,235 @@
+
+	private final Integer bulkProcessorFlushMaxActions;
+	private final Integer bulkProcessorFlushMaxSizeMb;
+	private final Integer bulkProcessorFlushIntervalMillis;
```
Why are you using boxed types here?

tzulitai replied:
I'll change this to `int` and use special values to represent that the user hasn't set a value.

rmetzger replied:
Ah, I see. You need boxed types for nullability. Then I would actually keep it as is.

tzulitai commented:
Thanks for the pointers. We'll include the profile `include-elasticsearch5` for Java 8 builds only.

rmetzger commented on flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/EmbeddedElasticsearchNodeEnvironment.java:
```diff
@@ -0,0 +1,37 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
This license header seems to be different from the other files.

rmetzger commented on flink-connectors/flink-connector-elasticsearch5/src/test/resources/log4j2.properties:
```diff
@@ -0,0 +1,27 @@
+# limitations under the License.
+################################################################################
+
+log4j.rootLogger=OFF, testlogger
```
Are you sure this is a valid log4j2 example?

Log4j2 files seem to look differently: http://howtodoinjava.com/log4j2/log4j-2-properties-file-configuration-example/

Is the ES5 connector pulling log4j2 as a dependency? Can we avoid that?

rmetzger commented on flink-connectors/flink-connector-elasticsearch5/src/test/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSinkITCase.java:
```diff
@@ -0,0 +1,68 @@
+	@Override
+	protected <T> ElasticsearchSinkBase<T> createElasticsearchSink(Map<String, String> userConfig,
+																List<InetSocketAddress> transportAddresses,
+																ElasticsearchSinkFunction<T> elasticsearchSinkFunction) {
```
Whitespace

(In general, this PR contains a lot of empty lines / vertical whitespace.

tzulitai replied:
Do you mean empty line? I don't think there are trailing whitespaces.
It isn't an empty line actually, the method parameters are at the right.
Need to drag the horizontal scroll to see it, otherwise on the IDE the seem to be aligned.

tzulitai replied:
I'll double-check the styling in this PR nevertheless.

rmetzger commented on flink-connectors/flink-connector-elasticsearch5/src/main/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchClientFactoryImpl.java:
```diff
@@ -0,0 +1,88 @@
+	/**
+	 * 	User-provided transport addresses.
+	 *
+	 * 	We are using {@link InetSocketAddress} because {@link TransportAddress} is not serializable in Elasticsearch 5.x.
```
Tab indentation

rmetzger commented on flink-connectors/flink-connector-elasticsearch5/pom.xml:
```diff
@@ -0,0 +1,121 @@
+			<groupId>org.apache.logging.log4j</groupId>
+			<artifactId>log4j-core</artifactId>
+			<version>2.7</version>
+		</dependency>
```
How does ES5 work when executed in a Flink program. Does it write the logs correctly into the taskmanager.log file using log4j2?

rmetzger commented on flink-connectors/flink-connector-elasticsearch2/src/test/java/org/apache/flink/streaming/connectors/elasticsearch2/ElasticsearchSinkITCase.java:
```diff
@@ -17,217 +17,51 @@
+	protected <T> ElasticsearchSinkBase<T> createElasticsearchSink(Map<String, String> userConfig,
+																List<InetSocketAddress> transportAddresses,
+																ElasticsearchSinkFunction<T> elasticsearchSinkFunction) {
+		return new ElasticsearchSink<>(userConfig, transportAddresses, elasticsearchSinkFunction);
```
Whitespace

tzulitai replied:
Same as above comment: there don't seem to be trailing spaces.
If you meant empty lines, there are actually method parameters on the far right.

rmetzger replied:
I mean the two newlines, which are empty

rmetzger commented on flink-connectors/pom.xml:
```diff
@@ -85,6 +86,17 @@ under the License.
+			to include it as part of Java 8 builds only.
+		-->
+		<profile>
+			<id>include-elasticsearch5</id>
```
Can you make this a profile that activates itself automatically when java 8 is available?
http://maven.apache.org/guides/introduction/introduction-to-profiles.html

rmetzger commented on .travis.yml:
```diff
@@ -16,14 +16,14 @@ matrix:
+      env: PROFILE="-Dhadoop.version=2.6.3 -Pinclude-yarn-tests,flink-fast-tests-a,include-kinesis,include-elasticsearch5 -Dmaven.javadoc.skip=true"
     - jdk: "oraclejdk8"
-      env: PROFILE="-Dhadoop.version=2.6.3 -Pinclude-yarn-tests,flink-fast-tests-b,include-kinesis -Dmaven.javadoc.skip=true"
+      env: PROFILE="-Dhadoop.version=2.6.3 -Pinclude-yarn-tests,flink-fast-tests-b,include-kinesis,include-elasticsearch5 -Dmaven.javadoc.skip=true"
```
With the automatic activation, you don't need these changes (see my other comment at the profile def)

tzulitai commented:
Hi @rmetzger, I've addressed your review comments. I still have some problems however, on the following:

1. I can't really reproduce the problem with the conflicts in `META-INF/services` dependencies, hence couldn't really try out whether that actually works correctly or if would actually use the `ServiceResourceTransformer` instead of `AppenderTransformer`. From the Maven documentation and what I understand from what the original ES2 documentation was trying to address, I think we can use the `ServiceResourceTransformer`. What do you think?

2. Regarding Log4j 2 dependency: That is required for the ES 5 Java client to log correctly, since the ES 5 Java API now uses Log4j 2 and does not detect logging implementations, must have a Log4j 2 API in the classpath. So the way I think the ES 5 connector is working is that the connector logs and Flink itself are logging using Log4j 1, while the internally used ES Java client is using Log4j 2 included exclusively in ES 5's POM.

I am still figuring out how to get the internal ES Java client in the ES 5 connector to log to TaskManager logs when using cluster execution, though (the connector log and Flink log is correctly logged, only the ES Java client log is missing).

I've included this `log4j2.properties` in a test project to be packaged for execution:
```
appender.file.type=File
appender.file.filename=${log.file}
appender.file.name=file
appender.file.layout.type=PatternLayout
appender.file.layout.pattern=%-4r [%t] %-5p %c %x - %m%n

rootLogger.level=info
rootLogger.appenderRef.file.ref=file
```

Somehow, it isn't picking the `log.file` property, which is set by the `flink-daemon.sh` script as a system property. Changing `log.file` to some another specific file path works.

rmetzger commented:
Gordon and I had a quick offline chat about this and decided the following:
1. We'll use the `ServiceResourceTransformer` because that's the correct way of solving the problem
2. There is a log4j2 to SLF4J bridge we'll try to use.

tzulitai commented:
@rmetzger The comments are all addressed now, including the last two problems.

fpompermaier commented on flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java:
```diff
@@ -0,0 +1,234 @@
+
+		clientFactory.cleanup();
+
+		if (hasFailure.get()) {
```
Do you think it could be possible to address also FLINK-5353 in this PR...?

rmetzger replied:
I would suggest to first merge this PR and then work on FLINK-5353 to not mix up too many things at the same time.

tzulitai commented:
Rebased to resolve conflicts in documentation changes.

mikedias commented on flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java:
```diff
@@ -0,0 +1,234 @@
+					if (response.hasFailures()) {
+						for (BulkItemResponse itemResp : response.getItems()) {
+							if (itemResp.isFailed()) {
+								LOG.error("Failed to index document in Elasticsearch: " + itemResp.getFailureMessage());
```
It's important to log the stacktrace as well: `LOG.error("message",  itemResp.getFailure().getCause())`

mikedias commented on flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java:
```diff
@@ -0,0 +1,234 @@
+						for (BulkItemResponse itemResp : response.getItems()) {
+							if (itemResp.isFailed()) {
+								LOG.error("Failed to index document in Elasticsearch: " + itemResp.getFailureMessage());
+								failureThrowable.compareAndSet(null, new RuntimeException(itemResp.getFailureMessage()));
```
Could be replaced by ` failureThrowable.compareAndSet(null, itemResp.getFailure().getCause());`

mikedias commented on flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java:
```diff
@@ -0,0 +1,234 @@
+
+				@Override
+				public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+					LOG.error(failure.getMessage());
```
It's important to log the stacktrace as well: `LOG.error(failure.getMessage(), failure);`

tzulitai commented:
Thank you for your comments @mikedias, I will address them.

tzulitai commented:
While rebasing #2861 on the restructured ES connectors in this PR, I jumped around different options on whether or not the `BulkProcessor` build process should be abstracted. In the end, I think it will still be best to keep that abstracted to keep duplicated code to the minimum.

To bridge the error logging problem, I introduced a `ElasticsearchApiCallBridge` interface to solve this. This call bridge can be further extended for other incompatible APIs that we bump into in the future.

tzulitai commented:
I think the PR is in a stable state for a final review now. I'll keep the PR as is until another review, sorry for jumping around as I was trying options for more future change proof.

fpompermaier commented:
Hi @tzulitai, congrats for the great work! For my use case it is important also to be resilient to malformed documents (https://issues.apache.org/jira/browse/FLINK-5353). Do you think you could address that issue as well?
That would be awesome for us :)

tzulitai commented:
Hi @fpompermaier! Thanks for bringing it up. However, as Robert suggested, lets not include that as this PR. But I'll like to discuss with you on the matter on the corresponding JIRA concurrently while this PR is in progress. I've left a comment on the JIRA already, lets move the discussion for this there :-)

rmetzger commented on docs/dev/connectors/elasticsearch.md:
```diff
@@ -23,158 +23,291 @@ specific language governing permissions and limitations
 
-Map<String, String> config = Maps.newHashMap();
+Map<String, String> config = new HashMap<>();
+config.put("cluster.name", "my-cluster-name")
```
There's a semikolon missing

rmetzger commented on docs/dev/connectors/elasticsearch.md:
```diff
@@ -23,158 +23,291 @@ specific language governing permissions and limitations
+DataStream<String> input = ...;
+
+Map<String, String> config = new HashMap<>();
+config.put("cluster.name", "my-cluster-name")
```
; missing

rmetzger commented on docs/dev/connectors/elasticsearch.md:
```diff
@@ -23,158 +23,291 @@ specific language governing permissions and limitations
 
-val config = new util.HashMap[String, String]
+val config = new java.util.HashMap[String, String]
+config.put("cluster.name", "my-cluster-name")
```
; again

tzulitai replied:
This is scala code, so the semicolon is ignored ;-) .

rmetzger commented:
+1 to merge.

I've tried the ES2 connector (just to check one of the connectors) and it worked well.

tzulitai commented:
Thank you for the review! I'll address the missing semicolons and merge this today.

tzulitai commented:
Merging to `master` ..
